### PR TITLE
Upgrade Go, controller-runtime and related dependencies

### PR DESCRIPTION
**Pull request by levan-m**
### What does this PR do?

See CECO-1114 for details about specific changes. 

* EDS
   v0.9.0-rc.2 -> v0.10.0-rc.4
* `sigs.k8s.io/controller-runtime` 
    v0.12.2 -> v0.16.4
* `k8s.io/api`, `k8s.io/apimachinery`, `k8s.io/client-go`
    0.24.2 -> v0.28.9
* `go.uber.org/zap` 
   v1.19.1 -> 0.25.0
* `go`
   1.19 -> 1.21

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Minimum Agent Versions

Are there minimum versions of the Datadog Agent and/or Cluster Agent required?

* Agent: vX.Y.Z
* Cluster Agent: vX.Y.Z

### Describe your test plan

Write there any instructions and details you may have to test your PR.

### Checklist

- [ ] PR has at least one valid label: `bug`, `enhancement`, `refactoring`, `documentation`, `tooling`, and/or `dependencies`
- [ ] PR has a milestone or the `qa/skip-qa` label

### CODE REVIEW

File: docs/configuration.v2alpha1.md [hunk 1]
```diff
@@ -340,7 +342,7 @@ In the table, `spec.override.nodeAgent.image.name` and `spec.override.nodeAgent.
...
 | [key].securityContext.seccompProfile.type | type indicates which kind of seccomp profile will be applied. Valid options are:  Localhost - a profile defined in a file on the node should be used. RuntimeDefault - the container runtime default profile should be used. Unconfined - no profile should be applied. |
-| [key].securityContext.supplementalGroups | A list of groups applied to the first process run in each container, in addition to the container's primary GID.  If unspecified, no groups will be added to any container. Note that this field cannot be set when spec.os.name is windows. |
+| [key].securityContext.supplementalGroups | A list of groups applied to the first process run in each container, in addition to the container's primary GID, the fsGroup (if specified), and group memberships defined in the container image for the uid of the container process. If unspecified, no additional groups are added to any container. Note that group memberships defined in the container image for the uid of the container process are still effective, even if they are not included in this list. Note that this field cannot be set when spec.os.name is windows. |
```
Comment (maycmlee) [inline docs/configuration.v2alpha1.md:345]:
```suggestion
| [key].securityContext.supplementalGroups | A list of groups applied to the first process run in each container, in addition to the container's primary GID, the fsGroup (if specified), and group memberships defined in the container image for the UID of the container process. If unspecified, no additional groups are added to any container. **Note**: Group memberships defined in the container image for the UID of the container process are still effective, even if they are not included in this list. This field cannot be set when spec.os.name is Windows. |
```

**Review by maycmlee — COMMENTED** [review]
A couple of suggestions

File: docs/configuration.v2alpha1.md (hunk 1 above)
Comment (levan-m) [reply docs/configuration.v2alpha1.md:345]: Some of these nested types get docs from the dependent packages and we can't really modify it. This should come from here
https://github.com/kubernetes/api/blob/02024d286e48d744890214fa862cb7ec9df865af/core/v1/types.go#L3785-L3793

**Comment by levan-m** [comment]
force-pushed after pulling changes from main (rebase) and follow-up commits to make build work.

**Review by fanny-jiang — APPROVED** [review]

**Review by fanny-jiang — APPROVED** [review]

**Review by fanny-jiang — APPROVED** [review]

**Pull request merged**